crates/aptos/src/governance/mod.rs

```diff
@@ -0,0 +1,185 @@
+use aptos_types::account_address::AccountAddress;
+use async_trait::async_trait;
+use clap::Parser;
+use reqwest::Url;
+use serde::Deserialize;
+use serde::Serialize;
+use std::fmt::Formatter;
+
+/// Tool for on-chain governance
+///
+#[derive(Parser)]
+pub enum GovernanceTool {
+    Propose(SubmitProposal),
+    Vote(SubmitVote),
+}
+
+impl GovernanceTool {
+    pub async fn execute(self) -> CliResult {
+        use GovernanceTool::*;
+        match self {
+            Propose(tool) => tool.execute_serialized().await,
+            Vote(tool) => tool.execute_serialized().await,
+        }
+    }
+}
+
+/// Submit proposal to other validators to be proposed on
+#[derive(Parser)]
+pub struct SubmitProposal {
+    #[clap(flatten)]
+    pub(crate) txn_options: TransactionOptions,
+    /// Delegated pool address to submit proposal on behalf of
+    #[clap(long)]
+    pub(crate) pool_address: AccountAddressWrapper,
+    /// Execution hash of the script to be voted on
+    #[clap(long, parse(try_from_str = read_hex_hash))]
+    pub(crate) execution_hash: HashValue,
+    /// Code location of the script to be voted on
+    #[clap(long)]
+    pub(crate) metadata_url: Url,
+
+    #[clap(flatten)]
+    pub(crate) prompt_options: PromptOptions,
+}
+
+#[async_trait]
+impl CliCommand<Transaction> for SubmitProposal {
+    fn command_name(&self) -> &'static str {
+        "SubmitProposal"
+    }
+
+    async fn execute(mut self) -> CliTypedResult<Transaction> {
+        // Validate the proposal metadata
+        let client = reqwest::ClientBuilder::default()
+            .tls_built_in_root_certs(true)
+            .build()
+            .map_err(|err| {
+                CliError::UnexpectedError(format!("Failed to build HTTP client {}", err))
+            })?;
+        let bytes = client
+            .get(self.metadata_url.clone())
+            .send()
+            .await
+            .map_err(|err| {
+                CliError::CommandArgumentError(format!(
+                    "Failed to fetch metadata url {}: {}",
+                    self.metadata_url, err
+                ))
+            })?
+            .bytes()
+            .await
+            .map_err(|err| {
+                CliError::CommandArgumentError(format!(
+                    "Failed to fetch metadata url {}: {}",
+                    self.metadata_url, err
+                ))
+            })?;
+        let metadata: ProposalMetadata = serde_json::from_slice(&bytes).map_err(|err| {
+            CliError::CommandArgumentError(format!(
+                "Metadata is not in a proper JSON format: {}",
+                err
+            ))
+        })?;
+        let metadata_hash = HashValue::sha3_256_of(&bytes);
+
+        println!("{}", metadata);
+        prompt_yes_with_override("Do you want to submit this proposal?", self.prompt_options)?;
+
+        // TODO: Allow Pool Address to use profile aliases
+        self.txn_options
+            .submit_script_function(
```

Comment:
Should we wait for when this transaction is executed and extract/display the proposal id?